### ISSUE TITLE
Is it necessary to perform layer replacement on te.xx? If not, is it effective to use te.fp8.autocast directly

### ISSUE THREAD

**Issue by wangli68** [NVIDIA/TransformerEngine#1556]
I did not replace te.xx, but directly replaced it with te.fp8.autocast at the mixed precision position (amp.autocast). Can fp8 be enabled for calculation？If I want to enable FP8 mixed precision calculation, how can I modify it?

for example：
class WanAttentionBlock(nn.Module):

    def __init__(self,
                 cross_attn_type,
                 dim,
                 ffn_dim,
                 num_heads,
                 window_size=(-1, -1),
                 qk_norm=True,
                 cross_attn_norm=False,
                 eps=1e-6):
        super().__init__()
        self.dim = dim
        self.ffn_dim = ffn_dim
        self.num_heads = num_heads
        self.window_size = window_size
        self.qk_norm = qk_norm
        self.cross_attn_norm = cross_attn_norm
        self.eps = eps

        # layers
        self.norm1 = WanLayerNorm(dim, eps)
        self.self_attn = WanSelfAttention(dim, num_heads, window_size, qk_norm,
                                           eps)
        self.norm3 = WanLayerNorm(
            dim, eps,
            elementwise_affine=True) if cross_attn_norm else nn.Identity()
        self.cross_attn = WANX_CROSSATTENTION_CLASSES[cross_attn_type](
            dim, num_heads, (-1, -1), qk_norm, eps)
        self.norm2 = WanLayerNorm(dim, eps)
        self.ffn = nn.Sequential(
            torch.nn.Linear(dim, ffn_dim), nn.GELU(approximate='tanh'),
            torch.nn.Linear(ffn_dim, dim))

        # modulation
        self.modulation = nn.Parameter(torch.randn(1, 6, dim) / dim**0.5)

    def forward(
        self,
        x,
        e,
        seq_lens,
        grid_sizes,
        freqs,
        context,
        context_lens,
    ):
        # assert e.dtype == torch.bfloat16
        # with amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
        with te.fp8_autocast(enabled=True,fp8_recipe=fp8_recipe):
            e = (self.modulation.to(dtype=e.dtype, device=e.device) + e).chunk(6, dim=1)
        # assert e[0].dtype == torch.bfloat16

        # self-attention
        y = self.self_attn(
            self.norm1(x).float() * (1 + e[1]) + e[0], seq_lens, grid_sizes,
            freqs)
        # with amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
        with te.fp8_autocast(enabled=True,fp8_recipe=fp8_recipe):
            x = x + y * e[2]

        # cross-attention & ffn function
        def cross_attn_ffn(x, context, context_lens, e):
            x = x + self.cross_attn(self.norm3(x), context, context_lens)
            y = self.ffn(self.norm2(x).float() * (1 + e[4]) + e[3])
            # with amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
            with te.fp8_autocast(enabled=True,fp8_recipe=fp8_recipe):
                x = x + y * e[5]
            return x

        x = cross_attn_ffn(x, context, context_lens, e)
        return x

**Comment by ksivaman** [comment]
`te.fp8_autocast` and torch AMP work differently and aren't exactly interchangeable (although they can be used together), and thus `te.fp8_autocast` is being used incorrectly in this script. `te.fp8_autocast` should be used to wrap the forward pass, and within that it will enable FP8 execution only for Transformer Engine layers. More details in our [getting started with FP8 docs](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/quickstart.html#Enabling-FP8). There are a few fixes needed in your script that you could try:

1. Replacing torch Linear with TE's Linear.
2. Wrapping the entire cross_attn_func in `te.fp8_autocast` .

Something like this:

```
def __init__(self,
             cross_attn_type,
             dim,
             ffn_dim,
             num_heads,
             window_size=(-1, -1),
             qk_norm=True,
             cross_attn_norm=False,
             eps=1e-6):
    super().__init__()
    self.dim = dim
    self.ffn_dim = ffn_dim
    self.num_heads = num_heads
    self.window_size = window_size
    self.qk_norm = qk_norm
    self.cross_attn_norm = cross_attn_norm
    self.eps = eps

    # layers
    self.norm1 = WanLayerNorm(dim, eps)
    self.self_attn = WanSelfAttention(dim, num_heads, window_size, qk_norm,
                                       eps)
    self.norm3 = WanLayerNorm(
        dim, eps,
        elementwise_affine=True) if cross_attn_norm else nn.Identity()
    self.cross_attn = WANX_CROSSATTENTION_CLASSES[cross_attn_type](
        dim, num_heads, (-1, -1), qk_norm, eps)
    self.norm2 = WanLayerNorm(dim, eps)
    self.ffn = nn.Sequential(
        te.Linear(dim, ffn_dim), nn.GELU(approximate='tanh'),
        te.Linear(ffn_dim, dim))

    # modulation
    self.modulation = nn.Parameter(torch.randn(1, 6, dim) / dim**0.5)

def forward(
    self,
    x,
    e,
    seq_lens,
    grid_sizes,
    freqs,
    context,
    context_lens,
):
    # assert e.dtype == torch.bfloat16
    # with amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
    with te.fp8_autocast(enabled=True,fp8_recipe=fp8_recipe):
        e = (self.modulation.to(dtype=e.dtype, device=e.device) + e).chunk(6, dim=1)
    # assert e[0].dtype == torch.bfloat16

    # self-attention
    y = self.self_attn(
        self.norm1(x).float() * (1 + e[1]) + e[0], seq_lens, grid_sizes,
        freqs)
    # with amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
    x = x + y * e[2]

    # cross-attention & ffn function
    def cross_attn_ffn(x, context, context_lens, e):
        x = x + self.cross_attn(self.norm3(x), context, context_lens)
        y = self.ffn(self.norm2(x).float() * (1 + e[4]) + e[3])
        # with amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
        x = x + y * e[5]
        return x
    
    with te.fp8_autocast(enabled=True,fp8_recipe=fp8_recipe):
        x = cross_attn_ffn(x, context, context_lens, e)
    return x
```

**Comment by sbhavani** [comment]
@wangli68 thanks for following up. To summarize: `te.fp8_autocast` / `te.autocast` only enables FP8 for Transformer Engine modules; it does not convert ordinary `torch.nn.Linear` or other PyTorch layers to FP8. For FP8 execution, replace eligible layers with TE modules such as `te.Linear` and wrap the forward region that calls them. In current TE, prefer `te.autocast` since `fp8_autocast` is deprecated.

For the LoRA concern: older PEFT versions reject `te.Linear`, which matches your traceback. PEFT added Transformer Engine LoRA support in huggingface/peft#3048, included in PEFT >= 0.19.0. Please upgrade PEFT or use a TE-aware LoRA wrapper.

Closing as resolved.
